packages/block-editor/src/components/block-binding-support/with-block-binding-support.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * External dependencies
+ */
+/**
+ * WordPress dependencies
+ */
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useEffect, useCallback, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import { useSelect } from '@wordpress/data';
+import { unlock } from '../../../../editor/src/lock-unlock';
+
+/**
+ * Conponent to bind an attribute to a prop.
+ *
+ * @param {Object}   props                   - The component props.
+ * @param {any}      props.attrValue         - The attribute value.
+ * @param {Function} props.onAttributeChange - The function to call when the attribute changes.
+ * @param {any}      props.propValue         - The prop value.
+ * @param {Function} props.onPropValueChange - The function to call when the prop value changes.
+ * @return {null} The component.
+ */
+const BlockBindingConnector = ( {
+	propValue,
+	onPropValueChange = () => {},
+
+	attrValue,
+	onAttributeChange,
+} ) => {
+	const lastPropValue = useRef();
+	const lastAttrValue = useRef();
+
+	useEffect( () => {
+		if ( propValue === lastPropValue.current ) {
+			return;
+		}
+
+		lastPropValue.current = propValue;
+		onPropValueChange( propValue );
+	}, [ onPropValueChange, propValue ] );
+
+	useEffect( () => {
+		if ( attrValue === lastAttrValue.current ) {
+			return;
+		}
+
+		lastAttrValue.current = attrValue;
+		onAttributeChange( attrValue );
+	}, [ onAttributeChange, attrValue ] );
+
+	return null;
+};
+
+const withBlockBindingSupport = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { attributes, name } = props;
+
+		const { getBlockBindingsSource } = unlock(
+			useSelect( blockEditorStore )
+		);
+
+		// Bail early if there are no bindings.
+		const bindings = attributes?.metadata?.bindings;
+		if ( ! bindings ) {
+			return <BlockEdit { ...props } />;
+		}
+
+		const BindingConnectorInstances = [];
+
+		Object.entries( bindings ).forEach( ( [ attrName, settings ], i ) => {
+			const source = getBlockBindingsSource( settings.source );
+
+			if ( source ) {
+				const { useSource } = source;
+				const attrValue = attributes[ attrName ];
+
+				/*
+				 * Pick the prop value and setter
+				 * from the source custom hook.
+				 */
+				const { value, updateValue } = useSource(
```

Comment:
We'll want to avoid calling `useSource()` conditionally (originally pointed out
in https://github.com/WordPress/gutenberg/pull/58085#discussion_r1468448826). This is something I
explored in https://github.com/WordPress/gutenberg/pull/58860

Reply:
One more thing - I see that you're not making use of the `placeholder` returned
from `useSource()` anymore. Why is that?

Reply:
yes, that would be the straightforward solution. It will make the implementation simpler and clearer, too.

Reply:
> One more thing - I see that you're not making use of the placeholder returned
from useSource() anymore. Why is that?

This was part of the refactoring part. Especially because it is not necessary in the challenge we are facing right now. Let me restore that part.
I'll remove the core/entity source handler, too.

Reply:
> explored in #58860

Same idea, right?

Reply:
> I think that we could just pass `settings.args` as a prop to
> `<BlockBindingConnector>` and then call `useSource()` hook inside of `<BlockBindingConnector>`

in that case we should pass the hook like a component prop?

Reply:
Yeah, that was my idea 🙂